datastream-common/src/main/java/com/linkedin/datastream/common/JsonUtils.java

```diff
@@ -96,6 +96,46 @@
     return object;
   }
 
+  /**
+   * Deserialize a JSON string into an object based on a type reference.
+   * This method allows the caller to specify precisely the desired output
+   * type for the target object.
+   * @param json JSON string
+   * @param typeRef type reference of the target object
+   * @param  mapper the ObjectMapper to use
+   * @return deserialized Java object
+   */
+  public static <T> T fromJson(String json, TypeReference<T> typeRef, ObjectMapper mapper) {
+    Validate.notNull(json, "null JSON string");
+    Validate.notNull(typeRef, "null type reference");
+    T object = null;
+    try {
+      object = mapper.readValue(json, typeRef);
+    } catch (IOException e) {
+      String errorMessage = "Failed to parse json: " + json;
+      ErrorLogger.logAndThrowDatastreamRuntimeException(LOG, errorMessage, e);
+    }
+    return object;
+  }
+
+  /**
+   * Serialize a Java object into JSON string.
+   * @param object object to be serialized
+   * @param  mapper the ObjectMapper to use
+   * @return JSON string
+   */
+  public static <T> String toJson(T object, ObjectMapper mapper) {
+    Validate.notNull(object, "null input object");
+    String out = null;
+    try {
+      out = mapper.writeValueAsString(object);
+    } catch (IOException e) {
+      String errorMessage = "Failed to serialize object: " + object;
+      ErrorLogger.logAndThrowDatastreamRuntimeException(LOG, errorMessage, e);
+    }
+    return out;
+  }
```

Comment:
Could you, please, have the existing `fromJson()` and `toJson()` invoke these methods after reconciling all differences, e.g.

```java
  public static <T> T fromJson(String json, TypeReference<T> typeRef) {
    return fromJson(json, typeRef, MAPPER);
  }

  public static <T> T fromJson(String json, TypeReference<T> typeRef, ObjectMapper mapper) {
    Validate.notNull(json, "null JSON string");
    Validate.notNull(typeRef, "null type reference");
    T object = null;
    try {
      object = mapper.readValue(json, typeRef);
    } catch (IOException e) {
      String errorMessage = "Failed to parse json: " + json;
      ErrorLogger.logAndThrowDatastreamRuntimeException(LOG, errorMessage, e);
    }
    return object;
  }

  public static <T> String toJson(T object) {
    return toJson(object, MAPPER);
  }

  public static <T> String toJson(T object, ObjectMapper mapper) {
    Validate.notNull(object, "null input object");
    StringWriter out = new StringWriter();   // Modified to match the old toJson()
    try {
      mapper.writeValue(out, object);        // ditto
    } catch (IOException e) {
      String errorMessage = "Failed to serialize object: " + object;
      ErrorLogger.logAndThrowDatastreamRuntimeException(LOG, errorMessage, e);
    }
    return out.toString();
  }
```

Reply:
Good point. I should have done this at the start. I cannot combine the ones that takes class and typereference. But the rest is doable.